backend/shared/shared/audit_log/viewsets.py

```diff
@@ -0,0 +1,104 @@
+from copy import copy
+from typing import Optional, Union
+
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import AnonymousUser
+from django.db.models import Model
+from django.db.models.base import ModelBase
+from rest_framework.viewsets import ModelViewSet
+from shared.audit_log import audit_logging
+from shared.audit_log.enums import Operation, Status
+
+User = get_user_model()
+
+
+class AuditLoggingModelViewSet(ModelViewSet):
+    method_to_operation = {
+        "POST": Operation.CREATE,
+        "GET": Operation.READ,
+        "PUT": Operation.UPDATE,
+        "PATCH": Operation.UPDATE,
+        "DELETE": Operation.DELETE,
+    }
+
+    def permission_denied(self, request, message=None, code=None):
+        audit_logging.log(
+            self._get_actor(),
+            self._get_operation(),
+            self._get_target(),
+            Status.FORBIDDEN,
+            ip_address=self._get_ip_address(),
+        )
+        super().permission_denied(request, message, code)
+
+    def retrieve(self, request, *args, **kwargs):
+        response = super().retrieve(request, *args, **kwargs)
+        audit_logging.log(
+            self._get_actor(),
+            Operation.READ,
+            self._get_target(),
+            ip_address=self._get_ip_address(),
+        )
+        return response
+
+    def perform_create(self, serializer):
+        super().perform_create(serializer)
+        audit_logging.log(
+            self._get_actor(),
+            Operation.CREATE,
+            serializer.instance,
+            ip_address=self._get_ip_address(),
+        )
+
+    def perform_update(self, serializer):
+        target_status_before = None
+        target_status_after = None
+
+        if hasattr(serializer.instance, "status"):
+            target_status_before = serializer.instance.status
+        super().perform_update(serializer)
+        if hasattr(serializer.instance, "status"):
+            target_status_after = serializer.instance.status
```

Comment:
Suggestion - implement if you feel like it: Only log status if target_status_before != target_status_after

Reply:
Sounds good, I'll add!